src/DynamoCoreWpf/Views/Core/WorkspaceView.xaml.cs

```diff
@@ -649,9 +653,92 @@ void vm_CurrentOffsetChanged(object sender, EventArgs e)
 
         void vm_ZoomChanged(object sender, EventArgs e)
         {
-            zoomBorder.SetZoom((e as ZoomEventArgs).Zoom);
-            if (PortContextMenu.IsOpen) DestroyPortContextMenu();
+            var newZoomScale = (e as ZoomEventArgs).Zoom;
+            zoomBorder.SetZoom(newZoomScale);
+            if (PortContextMenu.IsOpen)
+            {
+                DestroyPortContextMenu();
+            }
+
+            CheckZoomScaleAndApplyNodeViewCache(newZoomScale);
+        }
+
+        #region NodeView_BitmapCache
+        private void CheckZoomScaleAndApplyNodeViewCache(double newZoomScale)
+        {
+            //disable bitmap caching if max zoom scale set to 0, or feature flag was unable to fetch;
+            if (ViewModel.MaxZoomScaleForBitmapCache == 0) return;
+
+            if (!ViewModel.StopNodeViewOpacityAnimations)
+            {
+                if (currentRenderScale > 0) // number of nodes reduced below max threshold
+                {
+                    Dispatcher.BeginInvoke(ClearNodeViewCache, DispatcherPriority.Normal);
+                }
+
+                return;
+            }
+
+            if (newZoomScale > ViewModel.MaxZoomScaleForBitmapCache)
+            {
+                if (currentRenderScale > 0)
+                {
+                    Dispatcher.BeginInvoke(ClearNodeViewCache, DispatcherPriority.Normal);
+                }
+                
+                return;
+            }
+
+            var newRenderScale = newZoomScale <= .2 ? .2 : newZoomScale <= .5 ? .5 : 1;
+
+            if (Math.Abs(newRenderScale - currentRenderScale) <= 0.01)
+            {
+                return;
+            }
+
+            currentRenderScale = newRenderScale;
+            Dispatcher.BeginInvoke(UpdateNodeViewCacheScale, DispatcherPriority.Normal);
+        }
+
+        /// <summary>
+        /// Clear the node view bitmap cache
+        /// </summary>
+
+        private void ClearNodeViewCache()
+        {
+            var nodes = this.ChildrenOfType<NodeView>();
+            foreach (var node in nodes)
+            {
+                node.CacheMode = null;
+            }
+
+            currentRenderScale = -1;
+        }
+
+        /// <summary>
+        /// Update the node view bitmap cache scale
+        /// </summary>
+        private void UpdateNodeViewCacheScale()
+        {
+            var nodes = this.ChildrenOfType<NodeView>();
+            BitmapCache sharedCache = null;
+            foreach (var node in nodes)
+            {
+                if (node.CacheMode is BitmapCache cache)
```

Comment:
is this just a null check?

Reply:
yes

Reply:
it might be more clear if you just check it's null.

Reply:
We need to cast anyway if the condition is true, I think it is better since it checks for null and also gives us the appropriate object